tests/integration/ie-tests.sh

```diff
@@ -115,47 +101,85 @@ test_pyspark() {
+  echo -e "$(cat test-pyspark.py | spark-client.pyspark --username=${USERNAME} --namespace ${NAMESPACE} --conf spark.executor.instances=2)" > pyspark.out
   cat pyspark.out
   pi=$(cat pyspark.out  | grep "^Pi is roughly" | rev | cut -d' ' -f1 | rev | cut -c 1-3)
   echo -e "Pyspark Pi Job Output: \n ${pi}"
-  spark-client.service-account-registry delete --username=ie-test
   rm test-pyspark.py pyspark.out
   if [ "${pi}" != "3.1" ]; then
       exit 1
   fi
 }
 
+test_pyspark() {
+  run_pyspark tests spark
+}
+
 test_restricted_account() {
-  spark-client.service-account-registry delete --username=spark --namespace tests
 
-  kubectl delete namespace tests
+  kubectl config set-context spark-context --namespace=tests --cluster=prod --user=spark
 
-  kubectl create namespace tests
+  run_example_job tests spark
+}
 
-  spark-client.service-account-registry create --context microk8s --username spark --namespace tests
+setup_user() {
+  USERNAME=$1
+  NAMESPACE=$2
 
-  kubectl config set-context spark-context --namespace=tests --cluster=prod --user=spark
+  kubectl create namespace ${NAMESPACE}
 
-  run_example_job tests spark
+  if [ "$#" -gt 2 ]
+  then
+    CONTEXT=$3
+    spark-client.service-account-registry create --context ${CONTEXT} --username ${USERNAME} --namespace ${NAMESPACE}
+  else
+    spark-client.service-account-registry create --username ${USERNAME} --namespace ${NAMESPACE}
+  fi
+
+}
+
+setup_user_admin_context() {
+  setup_user spark tests
+}
+
+setup_user_restricted_context() {
+  setup_user spark tests microk8s
+}
+
+cleanup_user() {
+  EXIT_CODE=$1
+  USERNAME=$2
+  NAMESPACE=$3
+
+  spark-client.service-account-registry delete --username=${USERNAME} --namespace ${NAMESPACE}
 
-  spark-client.service-account-registry delete --username=spark --namespace tests
+  account_not_found_counter=$(spark-client.service-account-registry get-conf --username=${USERNAME} --namespace ${NAMESPACE} 2>&1 | grep -c NotFound)
 
-  kubectl delete namespace tests
+  if [ "${account_not_found_counter}" == "0" ]; then
+      exit 2
+  fi
 
-  account_deleted=$(spark-client.service-account-registry get-conf --username=spark --namespace tests 2>&1 | grep -c NotFound)
+  kubectl delete namespace ${NAMESPACE}
 
-  if [ "${account_deleted}" == "0" ]; then
+  if [ "${EXIT_CODE}" != "0" ]; then
       exit 1
   fi
 }
 
+cleanup_user_success() {
+  cleanup_user 0 spark tests
+}
+
+cleanup_user_failure() {
+  cleanup_user 1 spark tests
+}
+
 setup_tests
 
-test_example_job
+(setup_user_admin_context && test_example_job && cleanup_user_success) || cleanup_user_failure
 
-test_spark_shell
+(setup_user_admin_context && test_spark_shell && cleanup_user_success) || cleanup_user_failure
 
-test_pyspark
+(setup_user_admin_context && test_pyspark && cleanup_user_success) || cleanup_user_failure
 
-test_restricted_account
\ No newline at end of file
+(setup_user_restricted_context && test_restricted_account && cleanup_user_success) || cleanup_user_failure
```

Comment:
I REALLY LOVE this! Very neat! Much better than before and than what I had in mind. Having tests in one liner with very create structure "prepare stage => test => cleanup (rollaback for failure) is absolutely gold! Well done!

Reply:
Thanks Enrico! :)